doc/fpmsyncd/hld_fpmsyncd-NTT.md

```diff
@@ -0,0 +1,320 @@
+
+|  Rev  |     Date     |                       Author                       | Change Description                                          |
+| :---: | :----------: | :------------------------------------------------: | ----------------------------------------------------------- |
+|  0.1  | Jul 14, 2023 | Kanji Nakano, Kentaro Ebisawa, Hitoshi Irino (NTT) | Initial version                                             |
+|  0.2  | Jul 30, 2023 |               Kentaro Ebisawa (NTT)                | Remove description about VRF which is not nessesary for NHG. Add High Level Architecture diagram. Add note related to libnl, Routing WG. Fix typo and improve explanations.  |
+
+### Scope  
+
+This document details the design and implementation of the "fpmsyncd extension" related to NextHop Group behavior in SONiC.
+The goal of this "fpmsyncd extension" is to integrate NextHop Group functionality into SONiC by writing NextHop Group entry from `fpmsyncd` to `APPL_DB` for NextHop Group operation in SONiC.
+
+### Overview 
+
+SONIC system has support for programming routes using the NextHop Group feature through the NextHop Group table in `APPL_DB` database.
+The idea is to have a more efficient system that would involve managing the NextHop Group in use by the route table separately, and simply have the route table specify a reference to which NextHop Group to use.
+Since at scale many routes will use the same NextHop Groups, this requires much smaller occupancy per route, and so more efficient building, transmission and parsing of per-route information. 
+
+The current version of `fpmsyncd` has no support to handle the NextHop Group netlink messages sent by zebra process when it uses the new `dplane_fpm_nl` module.
+This implementation modifies the `fpmsyncd` code to handle `RTM_NEWNEXTHOP` and `RTM_DELNEXTHOP` events and write it to the database.
+Also, the `fpmsyncd` was modified to use the NextHop Group ID (`nexthop_group`) when programming the route to the `ROUTE_TABLE`.
+
+These capabilities consist in:
+- `fpmsyncd` is responsible for enabling the SET/DEL setting of `NEXTHOP_GROUP_TABLE` for `APPL_DB` in Redis DB.
+
+### Requirements
+
+`Fpmsyncd extension` requires:
+- `fpmsyncd` to handle `RTM_NEWNEXTHOP` and `RTM_DELNEXTHOP` events from zebra via `dplane_fpm_nl`
+- `fpmsyncd` to SET/DEL routes to `APPL_DB: ROUTE_TABLE` using `nexthop_group`
+- `fpmsyncd` to SET/DEL NextHop Group entry to `APPL_DB: NEXTHOP_GROUP_TABLE`
+
+### Architecture Design 
+<!--
+
+This section covers the changes that are required in the SONiC architecture. In general, it is expected that the current architecture is not changed.
+This section should explain how the new feature/enhancement (module/sub-module) fits in the existing architecture. 
+
+If this feature is a SONiC Application Extension mention which changes (if any) needed in the Application Extension infrastructure to support new feature.
+-->
+
+This design modifies `fpmsyncd` to use the new `APPL_DB` tables.
+
```

Comment:
As discussed during the review, it is more about move of logic of NHG creation from SWSS (Route/NHG OA) to Zebra. What is the main motivation for this move ? How such features as Dynamically Ordered ECMP (used by MSFT) with per NHGm's Sequence number work with this move ?

Reply:
> What is the main motivation for this move ? 

Motivation is to reduce NHG/ROUTE translation in zebra.
Also, supporting NHG in fpmsyncd as a first step to improve update performance (reducing DB change) by zebra to update a NHG shared among routes instead of updating all ECMP routes in ROUTE entry. (e.g., when NHG member was reduced from n to n-1)

> How such features as Dynamically Ordered ECMP (used by MSFT) with per NHGm's Sequence number work with this move ?

My understanding is this would not impact people using SWSS to create NHG if the feature is disabled (by default) since fpm will not send `RTM_NEWNEXTHOP` and thus fpmsyncd will set entry in CONFIG_DB using ROUTE entry as today.
Do you expect anyone to use Dynamically Ordered ECMP / per NHGm's Sequence number while enabling NHG in fpmsyncd in near future? (While I agree that it would be nice to have both available in long term but could be a different PR in future)

Reply:
The use case would be BGP PIC, and recursive routes handling. BGP PIC has started in design discussion in routing WG. Recursive routes support would be discussed after.

https://lists.sonicfoundation.dev/g/sonic-wg-routing/wiki/34786

Reply:
I have updated "Scope" for clarification and added PIC example in "Overview" section.